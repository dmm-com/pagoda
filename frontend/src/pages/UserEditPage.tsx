import { zodResolver } from "@hookform/resolvers/zod/dist/zod";
import { Box, Typography, Button, Container } from "@mui/material";
import { useSnackbar } from "notistack";
import React, { FC, useEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { Link, useNavigate } from "react-router-dom";
import { useToggle } from "react-use";

import { AironeBreadcrumbs } from "components/common/AironeBreadcrumbs";
import { Confirmable } from "components/common/Confirmable";
import { Loading } from "components/common/Loading";
import { PageHeader } from "components/common/PageHeader";
import { UserForm } from "components/user/UserForm";
import { UserPasswordFormModal } from "components/user/UserPasswordFormModal";
import { schema, Schema } from "components/user/userForm/UserFormSchema";
import { useAsyncWithThrow } from "hooks/useAsyncWithThrow";
import { useFormNotification } from "hooks/useFormNotification";
import { usePrompt } from "hooks/usePrompt";
import { useTypedParams } from "hooks/useTypedParams";
import { aironeApiClient } from "repository/AironeApiClient";
import { topPath, usersPath } from "routes/Routes";
import {
  extractAPIException,
  isResponseError,
} from "services/AironeAPIErrorUtil";
import { ServerContext } from "services/ServerContext";

export const UserEditPage: FC = () => {
  const { userId } = useTypedParams<{ userId?: number }>();
  const willCreate = userId == null;

  const navigate = useNavigate();
  const { enqueueSnackbar } = useSnackbar();
  const { enqueueSubmitResult } = useFormNotification("ユーザ", willCreate);
  const [shouldRefresh, toggleShouldRefresh] = useToggle(false);

  const {
    formState: { isValid, isDirty, isSubmitting, isSubmitSuccessful },
    handleSubmit,
    reset,
    setError,
    control,
  } = useForm<Schema>({
    resolver: zodResolver(schema),
    mode: "onBlur",
  });

  usePrompt(
    isDirty && !isSubmitSuccessful,
    "編集した内容は失われてしまいますが、このページを離れてもよろしいですか？"
  );

  const user = useAsyncWithThrow(async () => {
    if (userId) {
      return await aironeApiClient.getUser(userId);
    }
  }, [userId, shouldRefresh]);

  useEffect(() => {
    !user.loading && reset(user.value);
  }, [user.value]);

  useEffect(() => {
    isSubmitSuccessful && navigate(usersPath());
  }, [isSubmitSuccessful]);

  // These state variables and handlers are used for password reset feature
  const [openModal, setOpenModal] = useState(false);
  const handleOpenModal = () => {
    setOpenModal(true);
  };
  const handleCloseModal = () => {
    setOpenModal(false);
  };

  const isCreateMode = useMemo(() => {
    return user.value?.id == null;
  }, [user.value]);

  const [isSuperuser, isMyself] = useMemo(() => {
    const serverContext = ServerContext.getInstance();
    return [
      serverContext?.user?.isSuperuser != null &&
        serverContext.user.isSuperuser,
      user.value?.id != null &&
        serverContext?.user?.id != null &&
        user.value.id === serverContext.user.id,
    ];
  }, [user.loading]);

  const handleSubmitOnValid = async (user: Schema) => {
    try {
      if (isCreateMode) {
        await aironeApiClient.createUser(
          user.username,
          user.password ?? "",
          user.email,
          user.isSuperuser
        );
      } else {
        await aironeApiClient.updateUser(
          userId ?? 0,
          user.username,
          user.email,
          user.isSuperuser
        );
      }
      enqueueSubmitResult(true);
    } catch (e) {
      if (e instanceof Error && isResponseError(e)) {
        await extractAPIException<Schema>(
          e,
          (message) => enqueueSubmitResult(false, `詳細: "${message}"`),
          (name, message) => {
            setError(name, { type: "custom", message: message });
            enqueueSubmitResult(false);
          }
        );
      } else {
        enqueueSubmitResult(false);
      }
    }
  };

  const handleCancel = () => {
    navigate(usersPath());
  };

  const handleRefreshToken = async () => {
    try {
      await aironeApiClient.updateUserToken();
      toggleShouldRefresh();
    } catch (e) {
      if (e instanceof Response) {
        const json = await e.json();
        const reason = json["code"];
        enqueueSnackbar(`Token の更新に失敗しました。詳細: ${reason}`, {
          variant: "error",
        });
      } else {
        enqueueSnackbar(`Token の更新に失敗しました。`, {
          variant: "error",
        });
      }
    }
  };

  return (
    <Box>
      <AironeBreadcrumbs>
        <Typography component={Link} to={topPath()}>
          Top
        </Typography>
        <Typography component={Link} to={usersPath()}>
          ユーザ管理
        </Typography>
        <Typography color="textPrimary">ユーザ情報の設定</Typography>
      </AironeBreadcrumbs>
      <PageHeader
        title={user.value != null ? user.value.username : "新規ユーザの作成"}
        description={user.value != null ? "ユーザ編集" : undefined}
      >
        <Box display="flex" justifyContent="center">
          <Box mx="4px">
            <Button
              variant="contained"
              color="info"
              disabled={isCreateMode || !(isMyself || isSuperuser)}
              onClick={handleOpenModal}
            >
              パスワードの再設定
            </Button>
            <UserPasswordFormModal
              userId={user.value?.id ?? 0}
              openModal={openModal}
              onClose={handleCloseModal}
            />
          </Box>
          <Box mx="4px">
            <Confirmable
              componentGenerator={(handleOpen) => (
                <Button
                  variant="contained"
                  color="info"
                  disabled={isCreateMode || !isMyself}
                  onClick={handleOpen}
                >
                  Access Token をリフレッシュ
                </Button>
              )}
              dialogTitle="AccessTokenを更新してもよろしいですか？ ※現在入力中の項目はリセットされます"
              onClickYes={() => handleRefreshToken()}
            />
          </Box>
        </Box>
      </PageHeader>

      {user.loading ? (
        <Loading />
      ) : (
        <Container>
          <UserForm
            user={user.value}
            control={control}
            isCreateMode={isCreateMode}
            isSuperuser={isSuperuser}
            isMyself={isMyself}
            isSubmittable={
              isDirty && isValid && !isSubmitting && !isSubmitSuccessful
            }
            handleSubmit={handleSubmit(handleSubmitOnValid)}
            handleCancel={handleCancel}
          />
        </Container>
      )}
    </Box>
  );
};
