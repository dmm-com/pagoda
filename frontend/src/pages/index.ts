// FIXME rethink export scope

export * from "./EntryCopyPage";
export * from "./EntryEditPage";
export * from "./EntryEditPage";
export * from "./EntryListPage";

export * from "./UnavailableErrorPage";
