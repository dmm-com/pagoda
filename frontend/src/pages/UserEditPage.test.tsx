/**
 * @jest-environment jsdom
 */

import { render, screen, act, waitFor } from "@testing-library/react";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import React from "react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";

import { TestWrapperWithoutRoutes } from "TestWrapper";
import { UserEditPage } from "pages/UserEditPage";

const server = setupServer(
  // getUser
  http.get("http://localhost/role/api/v2/roles/1", () => {
    return HttpResponse.json({
      id: 1,
      username: "user1",
      email: "user1@example.com",
      is_superuser: false,
    });
  })
);

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

describe("EditUserPage", () => {
  Object.defineProperty(window, "django_context", {
    value: {
      user: {
        is_superuser: false,
      },
    },
    writable: false,
  });

  test("should match snapshot", async () => {
    const router = createMemoryRouter([
      {
        path: "/",
        element: <UserEditPage />,
      },
    ]);
    const result = await act(async () => {
      return render(<RouterProvider router={router} />, {
        wrapper: TestWrapperWithoutRoutes,
      });
    });
    await waitFor(() => {
      expect(screen.queryByTestId("loading")).not.toBeInTheDocument();
    });

    expect(result).toMatchSnapshot();
  });
});
