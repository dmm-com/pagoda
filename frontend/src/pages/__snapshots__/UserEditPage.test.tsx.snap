// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserPage should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-1baykhj"
        >
          <div
            class="MuiBox-root css-t0pdf1"
          >
            <div
              class="MuiBox-root css-jqffe8"
            >
              <nav
                aria-label="breadcrumb"
                class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
              >
                <ol
                  class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
                >
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/"
                    >
                      Top
                    </a>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/users"
                    >
                      ユーザ管理
                    </a>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
                    >
                      ユーザ情報の設定
                    </p>
                  </li>
                </ol>
              </nav>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1cs9h4l"
        >
          <div
            class="MuiBox-root css-1519avi"
          >
            <div
              class="MuiBox-root css-9rtbay"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2toz8d-MuiTypography-root"
                id="title"
              >
                新規ユーザの作成
              </h6>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
                id="description"
              />
              <div
                class="MuiBox-root css-hoapfw"
              >
                <div
                  class="MuiBox-root css-1l4w6pd"
                >
                  <div
                    class="MuiBox-root css-qbrse1"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-mnflth-MuiButtonBase-root-MuiButton-root"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      パスワードの再設定
                    </button>
                  </div>
                  <div
                    class="MuiBox-root css-qbrse1"
                  >
                    <div
                      class="MuiBox-root css-0"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-mnflth-MuiButtonBase-root-MuiButton-root"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        Access Token をリフレッシュ
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-lhgpb-MuiDivider-root"
            />
          </div>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-196rgkg"
            >
              <div
                class="MuiBox-root css-qbrse1"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-1mns5ui-MuiButtonBase-root-MuiButton-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  保存
                </button>
              </div>
              <div
                class="MuiBox-root css-qbrse1"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  キャンセル
                </button>
              </div>
            </div>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
              style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
            >
              <table
                class="MuiTable-root table table-bordered css-1xwxv7r-MuiTable-root"
              >
                <thead
                  class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head css-1wnroms-MuiTableRow-root"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-16h3rs8-MuiTableCell-root"
                      scope="col"
                    >
                      項目
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-16h3rs8-MuiTableCell-root"
                      scope="col"
                    >
                      内容
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                    >
                      名前
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                    >
                      <form
                        class="MuiBox-root css-nlbhye"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                        >
                          <div
                            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                              id=":r5:"
                              name="username"
                              placeholder="ユーザ名を入力してください"
                              type="text"
                              value=""
                            />
                            <fieldset
                              aria-hidden="true"
                              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                            >
                              <legend
                                class="css-w4cd9x"
                              >
                                <span
                                  aria-hidden="true"
                                  class="notranslate"
                                >
                                  ​
                                </span>
                              </legend>
                            </fieldset>
                          </div>
                        </div>
                      </form>
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                    >
                      メールアドレス
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                    >
                      <form
                        class="MuiBox-root css-nlbhye"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                        >
                          <div
                            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                              id=":r6:"
                              name="email"
                              placeholder="メールアドレスを入力してください"
                              type="email"
                              value=""
                            />
                            <fieldset
                              aria-hidden="true"
                              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                            >
                              <legend
                                class="css-w4cd9x"
                              >
                                <span
                                  aria-hidden="true"
                                  class="notranslate"
                                >
                                  ​
                                </span>
                              </legend>
                            </fieldset>
                          </div>
                        </div>
                      </form>
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                    >
                      パスワード
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                    >
                      <form
                        class="MuiBox-root css-nlbhye"
                      >
                        <div
                          class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                        >
                          <div
                            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                          >
                            <input
                              aria-invalid="false"
                              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                              id=":r7:"
                              name="password"
                              placeholder="パスワードを入力してください"
                              type="password"
                              value=""
                            />
                            <fieldset
                              aria-hidden="true"
                              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                            >
                              <legend
                                class="css-w4cd9x"
                              >
                                <span
                                  aria-hidden="true"
                                  class="notranslate"
                                >
                                  ​
                                </span>
                              </legend>
                            </fieldset>
                          </div>
                        </div>
                      </form>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1baykhj"
      >
        <div
          class="MuiBox-root css-t0pdf1"
        >
          <div
            class="MuiBox-root css-jqffe8"
          >
            <nav
              aria-label="breadcrumb"
              class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
            >
              <ol
                class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
              >
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                    data-discover="true"
                    href="/ui/"
                  >
                    Top
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                    data-discover="true"
                    href="/ui/users"
                  >
                    ユーザ管理
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
                  >
                    ユーザ情報の設定
                  </p>
                </li>
              </ol>
            </nav>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1cs9h4l"
      >
        <div
          class="MuiBox-root css-1519avi"
        >
          <div
            class="MuiBox-root css-9rtbay"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2toz8d-MuiTypography-root"
              id="title"
            >
              新規ユーザの作成
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              id="description"
            />
            <div
              class="MuiBox-root css-hoapfw"
            >
              <div
                class="MuiBox-root css-1l4w6pd"
              >
                <div
                  class="MuiBox-root css-qbrse1"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-mnflth-MuiButtonBase-root-MuiButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    パスワードの再設定
                  </button>
                </div>
                <div
                  class="MuiBox-root css-qbrse1"
                >
                  <div
                    class="MuiBox-root css-0"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-mnflth-MuiButtonBase-root-MuiButton-root"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      Access Token をリフレッシュ
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-lhgpb-MuiDivider-root"
          />
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-196rgkg"
          >
            <div
              class="MuiBox-root css-qbrse1"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-1mns5ui-MuiButtonBase-root-MuiButton-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                保存
              </button>
            </div>
            <div
              class="MuiBox-root css-qbrse1"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                キャンセル
              </button>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <table
              class="MuiTable-root table table-bordered css-1xwxv7r-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1wnroms-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-16h3rs8-MuiTableCell-root"
                    scope="col"
                  >
                    項目
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-16h3rs8-MuiTableCell-root"
                    scope="col"
                  >
                    内容
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                  >
                    名前
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                  >
                    <form
                      class="MuiBox-root css-nlbhye"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                            id=":r5:"
                            name="username"
                            placeholder="ユーザ名を入力してください"
                            type="text"
                            value=""
                          />
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-w4cd9x"
                            >
                              <span
                                aria-hidden="true"
                                class="notranslate"
                              >
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </form>
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                  >
                    メールアドレス
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                  >
                    <form
                      class="MuiBox-root css-nlbhye"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                            id=":r6:"
                            name="email"
                            placeholder="メールアドレスを入力してください"
                            type="email"
                            value=""
                          />
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-w4cd9x"
                            >
                              <span
                                aria-hidden="true"
                                class="notranslate"
                              >
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </form>
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root css-irfqwy-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1wgb810-MuiTableCell-root"
                  >
                    パスワード
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1qtwksv-MuiTableCell-root"
                  >
                    <form
                      class="MuiBox-root css-nlbhye"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root css-1myhl20-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                            id=":r7:"
                            name="password"
                            placeholder="パスワードを入力してください"
                            type="password"
                            value=""
                          />
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-w4cd9x"
                            >
                              <span
                                aria-hidden="true"
                                class="notranslate"
                              >
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </form>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
